framework/src/actions/SetupRecoverFileBaseAction.C

```diff
@@ -134,9 +135,26 @@ SetupRecoverFileBaseAction::getCheckpointFiles(std::set<std::string> & files)
     checkpoint_dirs.insert(FileOutput::getOutputFileBase(_app, "_out_cp"));
 
   // Add the directories from any existing checkpoint objects
-  const std::vector<Checkpoint *> ptrs = _app.getOutputWarehouse().getOutputs<Checkpoint>();
-  for (std::vector<Checkpoint *>::const_iterator it = ptrs.begin(); it != ptrs.end(); ++it)
-    checkpoint_dirs.insert((*it)->directory());
+  const std::vector<Action *> actions = _awh.getActionsByName("add_output");
+  for (std::vector<Action *>::const_iterator it = actions.begin(); it != actions.end(); ++it)
+  {
+    // Get the parameters from the MooseObjectAction
+    MooseObjectAction * moose_object_action = static_cast<MooseObjectAction *>(*it);
+    const InputParameters & params = moose_object_action->getObjectParams();
+
+    // Loop through the actions and add the necessary directories to the list to check
+    if (moose_object_action->getParam<std::string>("type") == "Checkpoint")
+    {
+      if (params.isParamValid("file_base"))
+        checkpoint_dirs.insert(common->getParam<std::string>("file_base") + "_cp");
+      else
+      {
+        std::ostringstream oss;
+        oss << "_" << (*it)->getShortName() << "_cp";
+        checkpoint_dirs.insert(FileOutput::getOutputFileBase(_app, oss.str()));
+      }
+    }
+  }
```

Comment:
I hate to be critical of both your patches but this a bit clunky too. This patch now loops through a list of unrelated actions inspecting the type to make a decision about something. Let's chat about alternative designs that don't have to interact with the Factory or other objects so much.